### DISCUSSION TITLE
LocalDateTime 역직렬화 문제

### DISCUSSION THREAD

**Discussion by YurimYang** [SCBJ-7/SCBJ-BE#147] [Q&A]
### 발생과정

- 야놀자 예약내역을 불러올 때 localDateTime의 문제로 internal Server Error가 발생했습니다.
![image](https://github.com/SCBJ-7/SCBJ-BE/assets/63856521/4283f098-bdfd-43fe-bfbc-d234f8689477)

### 원인

- 야놀자 예약내역을 response 해줄 때, @ResController로부터 리턴하는 값이 Json 스트링 형식으로 변환되어 응답되는데, 이를 ObjectMapper에 의해 직렬화 한다. (Object -> Json)
- 이때, ObjectMapper는 직렬화/역직렬화를 하는데 JavaTimeModule를 등록해야 LocalDateTime을 올바르게 처리하도록 할 수 있다.

### 해결방안
- ObjectMapper 
  - 직렬화 : Java 객체 -> Json
  - 역직렬화 : Json -> Java 객체
  - 이때, 복잡한 데이터 유형에 속하는 LocalDateTime을 적절하게 처리하기 위해서는 JavaTimeModule를 직접 추가해줘야한다.
 
- JavaTimeModule
   - JAVA8부터 도입된 날짜 및 시간 API를 Jackson 라이브러리에서 적절하게 처리할 수 있게 해주는 모듈 
   - 기본적으로 Jackson 라이브러리는 Java 8의 새로운 날짜 및 시간 타입을 인식하지 못해서 직렬화 및 역직렬화에서 문제가 발생할 수 있다
 
- 즉, LocalDateTime은 표준 Java 라이브러리 일부가 아니고, Java8부터 등장한 JSR310의 날짜/시간 API의 일부이므로 추가적인 등록 작업이 필요하다. 
```
//deserialize
implementation 'com.fasterxml.jackson.datatype:jackson-datatype-jsr310'
implementation 'com.fasterxml.jackson.core:jackson-databind'
```
```
    @Bean
    public ObjectMapper objectMapper() {
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.registerModule(new JavaTimeModule()); // JavaTimeModule 등록
        objectMapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
        return objectMapper;
    }
```
```
    @JsonSerialize(using = LocalDateTimeSerializer.class)
    @JsonDeserialize(using = LocalDateTimeDeserializer.class)
    @Column(nullable = false)
    @Comment("시작일")
    private LocalDateTime startDate;

    @JsonSerialize(using = LocalDateTimeSerializer.class)
    @JsonDeserialize(using = LocalDateTimeDeserializer.class)
    @Column(nullable = false)
    @Comment("종료일")
    private LocalDateTime endDate;
```

**Comment by Kwonminwoo** [comment]
해당 부분 잊고 있었는데 덕분에 기억이 났네요!!

**Reply by wocjf0513** [reply]
오오 이런게 있었군요 감사합니다!!
